addons/account/models/account_move.py

```diff
@@ -3074,30 +3074,28 @@ def _extend_with_attachments(self, attachments, new=False):
         """
         success = False
 
-        for file_data in attachments._unwrap_edi_attachments(): # sorted by priority
-
-            decoder = self._get_edi_decoder(file_data, new=new)
-
-            try:
-                if decoder and not success:
-                    with self.env.cr.savepoint(), self._get_edi_creation() as invoice:
-                        # pylint: disable=not-callable
-                        success = decoder(invoice, file_data, new)
+        # sorted by priority
+        for file_data in attachments._unwrap_edi_attachments():
+            if not success and (decoder := self._get_edi_decoder(file_data, new=new)):
+                try:
+                    with self.env.cr.savepoint():
+                        with self._get_edi_creation() as invoice:
+                            # pylint: disable=not-callable
+                            success = decoder(invoice, file_data, new)
                         if success:
```

Comment:
Pure refactoring with no purpose, right? If purely cosmetic, remove that diff

Reply:
No, this is a bugfix that is part of this refactor. We do this so that if there is an exception while completing the invoice using the received XML's data, the invoice gets saved (which happens on exiting `self._get_edi_creation()`) before we link it to the the purchase order.